src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -156,6 +201,492 @@ MultiRouterPlanCreate(Query *originalQuery, Query *query,
+
+	ExtractQueryWalker((Node *) query, &queryList);
+	foreach(queryCell, queryList)
+	{
+		Query *subquery = (Query *) lfirst(queryCell);
+
+		Assert(subquery->commandType == CMD_SELECT);
+
+		if (subquery->limitCount != NULL)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot perform distributed planning for the given "
+								   "modification"),
+							errdetail(
+								"LIMIT clauses are not allowed in INSERT ... SELECT "
+								"queries")));
+		}
+
+		if (subquery->limitOffset != NULL)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot perform distributed planning for the given "
+								   "modification"),
+							errdetail(
+								"OFFSET clauses are not allowed in INSERT ... SELECT "
+								"queries")
+
+							));
+		}
+
+		if (subquery->windowClause != NULL)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot perform distributed planning for the given "
+								   "modification"),
+							errdetail(
+								"Window functions are not allowed in INSERT ... SELECT "
+								"queries")));
+		}
+
+		if (subquery->setOperations != NULL)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot perform distributed planning for the given "
+								   "modification"),
+							errdetail(
+								"Union, Intersect and Except are currently unsupported are not allowed in INSERT ... SELECT "
+								"queries")));
+		}
+	}
+}
+
+
+/*
+ * ErrorIfNotAllParticipatingTablesAreColocated errors out of all tables
+ * referenced in the query are not colocated.
+ */
+static void
+ErrorIfNotAllParticipatingTablesAreColocated(Query *query)
+{
+	List *relationIdList = RelationIdList(query);
+	ListCell *relationIdCell = NULL;
+	uint64 colocationId = INVALID_COLOCATION_ID;
+	bool tablesAreColocated = true;
+
+	foreach(relationIdCell, relationIdList)
+	{
+		Oid relationId = lfirst_oid(relationIdCell);
+		uint64 currentColocationId = TableColocationId(relationId);
+
+		if (currentColocationId == INVALID_COLOCATION_ID)
+		{
+			tablesAreColocated = false;
+			break;
+		}
+
+		/* set for the first table */
+		if (colocationId == INVALID_COLOCATION_ID)
+		{
+			colocationId = currentColocationId;
+		}
+
+		if (colocationId != currentColocationId)
+		{
+			tablesAreColocated = false;
+			break;
+		}
+	}
+
+
+	if (!tablesAreColocated)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("all participating tables should be colocated")));
+	}
+}
+
+
+/*
+ * ErrorIfInsertPartitionColumnDoesNotMatchSelect checks whether the INSERTed table's
+ * partition column value matches with the any of the SELECTed table's partition column.
+ */
+static void
+ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query)
+{
+	ListCell *targetEntryCell = NULL;
+	uint32 rangeTableId = 1;
+	RangeTblEntry *insertRte = linitial(query->rtable);
+	RangeTblEntry *subqueryRte = lsecond(query->rtable);
```

Comment:
I think it's a bad idea to hardcode subquery/insert RTEs this way. We should really determine this once, and then pass those down.